### PR TITLE
`CodeBlock`: update to support dynamic content

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary
If merged, this PR:
 - updates the `CodeBlock` component to enable it to render updated dynamic content
 - addresses an issue where line numbers are misaligned due to [an issue with font declaration](https://github.com/hashicorp/design-system/pull/1853/commits/cc377c8076e6fca1ebbd24435e956a2514518809)
 - suppresses [a build warning](https://hashicorp.atlassian.net/browse/HDS-2860) caused by parsing Ember comments in the component template.

Limitations:
 - dynamic content does not work with `@hasLineNumbers={{true}}`; this comes from a limitation with the [line-numbers Prism plugin](https://prismjs.com/plugins/line-numbers/) where it can only be instantiated/called once.
 - `@hasLineNumbers={{true}}` does not work as expected when used in a `Tabs` tab that is not visible at render time.

**Showcase:** https://hds-showcase-git-test-codeblock-with-dynamic-content-hashicorp.vercel.app/components/code-block

### :camera_flash: Screenshots

Misalignment issue in HCP. The more line there are the more you can see the issue.
<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td>

01-hcp-line-before

</td><td>

02-hcp-line-after

</td></tr>
</table>

### :link: External links

- Related Jira ticket: [HDS-2827](https://hashicorp.atlassian.net/browse/HDS-2827)

***

### 👀 Component checklist

- [x] Percy was checked for any visual regression
- [x] A11y tests have been run locally (`yarn test:a11y --filter="COMPONENT-NAME"`)
- [x] If documenting a new component, an acceptance test that includes the `a11yAudit` has been added
- [x] A changelog entry was added via [Changesets](https://github.com/changesets/changesets) if needed (see [templates here](https://github.com/hashicorp/design-system/blob/main/wiki/Website-Changelog.md#templates-for-npm-packages))

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

[HDS-2827]: https://hashicorp.atlassian.net/browse/HDS-2827?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/addon/components/hds/code-block/index.js [hunk 1]
```diff
@@ -107,18 +107,20 @@ export default class HdsCodeBlockIndexComponent extends Component {
...
+          this.prismCode = htmlSafe(Prism.util.encode(code));
+        }
+
+        // Force prism-line-numbers plugin initialization, required for Prism.highlight usage
+        // See https://github.com/PrismJS/prism/issues/1234
+        Prism.hooks.run('complete', {
+          code,
+          element,
+        });
+      }, 100);
```
Comment (KristinLBradley) [inline packages/components/addon/components/hds/code-block/index.js:123]: I tried using `next` to replace this `setTimeout` and it still functions.

I tried using `schedule` instead and got errors but maybe I was using it wrong so I'll experiment a bit more.

Comment (didoo) [reply packages/components/addon/components/hds/code-block/index.js]: OK, then let's wait for @alex-ju to come back from PTO and we can then decide how to take it from here.

Comment (KristinLBradley) [reply packages/components/addon/components/hds/code-block/index.js]: It seems to work although I can do more testing.

Comment (alex-ju) [reply packages/components/addon/components/hds/code-block/index.js]: if we could also test this change to make sure it addresses the actual need in `boundary` (where you identified the issue) that would be even better – let me know if I can be of help

Comment (KristinLBradley) [reply packages/components/addon/components/hds/code-block/index.js]: @alex-ju I tried figuring out how to add an integration test but haven't been able to figure out a way to do it.

I can try to test it just in Boundary next though.

Comment (alex-ju) [reply packages/components/addon/components/hds/code-block/index.js]: cool! I can look into the integration test bit

Comment (KristinLBradley) [reply packages/components/addon/components/hds/code-block/index.js]: I tested the CodeBlock updates in Boundary and everything worked as expected. The content updates dynamically.

File: packages/components/tests/integration/components/hds/code-block/index-test.js [hunk 2]
```diff
@@ -33,6 +38,23 @@ module('Integration | Component | hds/code-block/index', function (hooks) {
...
+  test('it renders the passed in dynamic content', async function (assert) {
+    this.set('value', "console.log('Hello world');");
+    await render(hbs`
+      <Hds::CodeBlock @value={{this.value}} id="test-code-block" />
+    `);
+    assert
+      .dom('#test-code-block pre code')
+      .hasText("console.log('Hello world');");
+    this.set('value', "console.log('Lorem ipsum');");
+    await settled();
```
Comment (alex-ju) [inline packages/components/tests/integration/components/hds/code-block/index-test.js:52]: I initially tried with `rerender`, but it's too 'quick' (in a sense that it doesn't wait for the whole promise to end) so I ended up using `settled`.

File: packages/components/tests/dummy/app/templates/components/code-block.hbs [hunk 3]
```diff
@@ -77,6 +79,34 @@ console.log(`I am ${codeLang} code`);"
...
+  <Shw::Text::Body>
+    <label for="input">Change input value:</label>
+    <input id="input" type="text" value={{this.input}} {{on "input" this.updateInput}} />
+  </Shw::Text::Body>
+
+  <Hds::CodeBlock
+    @value={{this.codeValue}}
+    @language="javascript"
+    @hasCopyButton={{true}}
+    @hasLineNumbers={{false}}
```
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/code-block.hbs:101]: I noticed that if you set this to "true" (and probably we should) when the lines gets updated, the line numbers don't.

File: packages/components/addon/components/hds/code-block/index.js [hunk 4]
```diff
@@ -107,27 +108,29 @@ export default class HdsCodeBlockIndexComponent extends Component {
...
+        // See https://github.com/PrismJS/prism/issues/1234
+        Prism.hooks.run('complete', {
+          code,
+          element,
+        });
+
+        // Force prism-line-highlight plugin initialization
+        // Context: https://github.com/hashicorp/design-system/pull/1749#discussion_r1374288785
+        if (this.args.highlightLines) {
+          schedule('afterRender', () => {
```
Comment (didoo) [inline packages/components/addon/components/hds/code-block/index.js:128]: do we still need to invoke the "afterRender" even if it's inside a `next()` callback?

**Review by didoo — COMMENTED** [review]
Left a couple of comments.

File: packages/components/addon/components/hds/code-block/index.js (hunk 4 above)
Comment (alex-ju) [reply packages/components/addon/components/hds/code-block/index.js:128]: yes, unfortunately, we still need to delay the `prism-line-highlight` for as much as possible, and `afterRender` seems to be the 'latest' we can use.

Comment (didoo) [reply packages/components/addon/components/hds/code-block/index.js:128]: OK (maybe add a comment in the code explaining this detail?)

File: packages/components/tests/dummy/app/templates/components/code-block.hbs (hunk 3 above)
Comment (alex-ju) [reply packages/components/tests/dummy/app/templates/components/code-block.hbs:101]: I looked into this during the first week of January and today again. Unfortunately, I can not see a way in which the `@hasLineNumbers` feature would work with dynamic content.

For this reason, I suggest the pragmatic approach of fixing the immediate issues (I updated the PR description to highlight what this PR covers and what doesn't) to address current instances in HCP and Boundary then leave the `hasLineNumbers` issue for future evaluation (it may require us to fork the addon or move away from Prism altogether). If you find this approach acceptable, I can create a ticket to track this use case and update the CodeBlock guidance to call out the current limitations.

File: packages/components/addon/components/hds/code-block/index.js (hunk 4 above)
Comment (alex-ju) [reply packages/components/addon/components/hds/code-block/index.js:129]: Added a comment explaining the need for `afterRender`

**Review by didoo — APPROVED** [review]
LGTM 👍
